text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+constitutes a "mutable use" are not stated here.
+
+[mutating use method]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/enum.PlaceContext.html#method.is_mutating_use
+
+## Interaction with `trait` aliases
+
+Trait aliases cannot be implemented. As such, there is no concern about compatibility between the
+`impl` restriction and `trait` aliases.
+
+# Drawbacks
+
+- Additional syntax for macros to handle
+
+# Alternatives
+
+- `impl` and `mut` restrictions could be attributes, similar to `#[non_exhaustive]`.
+
+# Prior art
+
+- The [`readonly` crate] simulates immutable fields outside of the defining module. Types with this
+  attribute cannot define `Deref`, which can be limiting. Additionally, it applies to all fields and
+  within the defining crate. The advantages of native read-only fields relating to borrow checking
+  also do not apply when using this crate.
+- The ability to restrict implementations of a trait can be simulated by a public trait in a private
+  module. This has the disadvantage that the trait is no longer nameable by external users,
+  preventing its use as a generic bound. Current diagnostics, while technically correct, are
+  unhelpful to downstream users.
+- Various other languages have read-only fields, including C++, C#, Java, TypeScript, Kotlin, and
+  Swift.
+
+# Unresolved questions
+
+- Should unrestricted "restrictions" be allowed to be explicitly stated, or should this be banned?
+  If allowed, should it be a warning?
+  - Both `impl` and `mut` make no sense to have unrestricted, as that is the default behavior. An
+    "unnecessary" restriction lint that is warn-by-default would be ideal. This would allow macros
+    to emit more flexible syntax. This warning could also be used for `pub(self)`.
+- How will restrictions work with `macro_rules!` matchers? There is currently a `vis` matcher, but
+  it is likely unwise to add a new matcher for each restriction. Should the new syntax be included
+  as part of the `vis` matcher? If so, it is important to note that restrictions are not the same
+  everywhere, if they are valid at all.
+  - Suggestions welcome!
+- What is the interaction between stability and restrictions?
+  - Suggestion: Visibility is an inherent part of the item; restrictions should be as well. Metadata
+    can be added in the future indicating when an item had its restriction lifted, if applicable.
+    The design for this is left to the language team as necessary.
+
+# Future possibilities
+
+- Sealed/exhaustive traits could happen in the future. This has the ability to impact coherence,
+  such that other crates could rely on the fact that the list of implementations is exhaustive. As
+  traits would default to unsealed, this does not have be decided now.
+- Trait items could gain proper visibility and/or restrictions of their own. This would allow
+  private and/or defaulted trait items that cannot be overridden.
```

Comment:
This is a great step towards sealed/exhaustive traits ❤️ Hence, I think restrictions should have as strong stability guarantees as visibility does now.